Gifski/util.swift

```diff
@@ -1117,3 +1169,206 @@ extension CGRect {
+  }
+
+  /// Returns a new result, mapping any failure value using the given
+  /// transformation and unwrapping the produced result.
+  ///
+  /// - Parameter transform: A closure that takes the failure value of the
+  ///   instance.
+  /// - Returns: A `Result` instance, either from the closure or the previous
+  ///   `.success`.
+  public func flatMapError<NewFailure>(
+    _ transform: (Failure) -> Result<Success, NewFailure>
+  ) -> Result<Success, NewFailure> {
+    switch self {
+    case let .success(success):
+      return .success(success)
+    case let .failure(failure):
+      return transform(failure)
+    }
+  }
+
+  /// Returns the success value as a throwing expression.
+  ///
+  /// Use this method to retrieve the value of this result if it represents a
+  /// success, or to catch the value if it represents a failure.
+  ///
+  ///     let integerResult: Result<Int, Error> = .success(5)
+  ///     do {
+  ///         let value = try integerResult.get()
+  ///         print("The value is \(value).")
+  ///     } catch error {
+  ///         print("Error retrieving the value: \(error)")
+  ///     }
+  ///     // Prints "The value is 5."
+  ///
+  /// - Returns: The success value, if the instance represent a success.
+  /// - Throws: The failure value, if the instance represents a failure.
+  public func get() throws -> Success {
+    switch self {
+    case let .success(success):
+      return success
+    case let .failure(failure):
+      throw failure
+    }
+  }
+}
+
+extension Result: Equatable where Success: Equatable, Failure: Equatable {}
+extension Result: Hashable where Success: Hashable, Failure: Hashable {}
+
+// To be able to use it in places that already have a local result
+// TODO: Remove this when using Swift 5
+typealias CoreResult = Result
+
+
+public protocol CancellableError: Error {
+    /// Returns true if this Error represents a cancelled condition
+	var isCancelled: Bool { get }
+}
+
+public struct CancellationError: CancellableError {
+	public var isCancelled = true
+}
+
+extension Error {
+    public var isCancelled: Bool {
+        do {
+            throw self
+        } catch let error as CancellableError {
+            return error.isCancelled
+        } catch URLError.cancelled {
+            return true
+        } catch CocoaError.userCancelled {
+            return true
+        } catch {
+        #if os(macOS) || os(iOS) || os(tvOS)
+            let pair = { ($0.domain, $0.code) }(error as NSError)
+            return pair == ("SKErrorDomain", 2)
+        #else
+            return false
+        #endif
+        }
+    }
+}
+
+extension Result {
+	/**
+	```
+	switch result {
+	case .success(let value):
+		print(value)
+	case .failure where result.isCancelled:
+		print("Cancelled")
+	case .failure(let error):
+		print(error)
+	}
+	```
+	*/
+    public var isCancelled: Bool {
+        do {
+            _ = try get()
+			return false
+        } catch {
+            return error.isCancelled
+		}
```

Comment:
indentation seems a bit off here (mixed tabs and spaces...)

Reply:
Fixed